File: src/main/java/br/com/estimular/atendimento/service/messaging/consumer/AbstractConsumer.java
package br.com.estimular.atendimento.service.messaging.consumer;

import lombok.SneakyThrows;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.serialization.Deserializer;
import org.springframework.beans.factory.DisposableBean;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public abstract class AbstractConsumer<K, V> implements DisposableBean {

    private final AtomicBoolean running = new AtomicBoolean(true);
    private final Consumer<K, V> consumer;
    private final ExecutorService executorService;

    public AbstractConsumer(
            String bootstrapServer,
            String topico,
            Class<? extends Deserializer<K>> keyClassDeserializer,
            Class<? extends Deserializer<V>> valueClassDeserializer) {
        Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "resumo-group");
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyClassDeserializer.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueClassDeserializer.getName());
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

        consumer = new KafkaConsumer<>(props);
        consumer.subscribe(Collections.singletonList(topico));

        executorService = Executors.newSingleThreadExecutor();
        executorService.submit(this::listem);
    }

    @SneakyThrows
    private void listem() {
        try {
            while (running.get()) {
                consumer
                        .poll(Duration.ofMillis(100))
                        .forEach(this::receber);
            }
        } finally {
            consumer.close();
        }
    }

    protected abstract void receber(ConsumerRecord<K, V> record);

    @Override
    public void destroy() {
        running.set(false);
        if (executorService != null) {
            executorService.shutdown();
        }
    }
}

File: src/main/kotlin/com/clinica/atendimento/service/consumer/AbstractConsumer.kt
package com.clinica.atendimento.service.consumer

import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.serialization.Deserializer
import org.springframework.beans.factory.DisposableBean

import java.time.Duration
import java.util.Collections
import java.util.Properties
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

abstract class AbstractConsumer<K, V>(
    bootstrapServer: String,
    topico: String,
    keyClassDeserializer: Class<out Deserializer<K>>,
    valueClassDeserializer: Class<out Deserializer<V>>
) : DisposableBean {

    private val running = AtomicBoolean(true)
    private val consumer: Consumer<K, V>
    private val executorService: ExecutorService

    init {
        val props = Properties()
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer)
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "resumo-group")
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyClassDeserializer.name)
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueClassDeserializer.name)
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")

        consumer = KafkaConsumer(props)
        consumer.subscribe(Collections.singletonList(topico))

        executorService = Executors.newSingleThreadExecutor()
        executorService.submit { listem() }
    }

    private fun listem() {
        try {
            while (running.get()) {
                consumer
                    .poll(Duration.ofMillis(100))
                    .forEach { this.receber(it) }
            }
        } finally {
            consumer.close()
        }
    }

    protected abstract fun receber(record: ConsumerRecord<K, V>)

    override fun destroy() {
        running.set(false)
        executorService.shutdown()
    }
}